### ISSUE TITLE
Make HDF5 output format usable outside of pandas

### ISSUE THREAD

**Issue by speth** [Cantera/cantera#830]
The HDF5 output option added to the `SolutionArray` class currently uses the `pandas` library to write the HDF5 files. However, `pandas`-generated HDF5 files introduce an additional layer of abstraction beyond the native structure of HDF5 datasets which makes them difficult to use except with `pandas`. For example, saving a `SolutionArray` generates an HDF5 file with keys:
```
df/axis0
df/axis1
df/block0_items
df/block0_values
```
where `df` is the given name of the dataset, `df/axis0` contains the column names, `df/axis1` is a list of integers, `df/block0_items` contains the column names again, and `df/block0_values` contains the 2D array of actual values.

Using a simpler, more direct structure would allow these files to be easily used with other libraries in Python (e.g. `h5py` and `pytables`) as well as in other languages such as Matlab or C++. For example, this could simply be
```
df/T
df/density
df/Y_H2
df/Y_H
...
```
where each dataset was the corresponding array of data. One question is whether it would make sense to store `df/Y` as a 2D array with an additional array for the species names. That would make it easier to do some operations when working with the same mechanism file, but make loading data for a different mechanism (with different species) more difficult.

Since the HDF5 output is new for version 2.5, we can still make changes without having to worry about backwards compatibility, and I think we should do so while we can.

This also applies to the HDF5 output proposed in #805.

**Comment by ischoegl** [comment]
I'm supportive of this discussion - the current HDF output makes use of `SolutionArray` objects which are currently only supported in Python. Regardless, it would be straight-forward to write compatible output formats that can be easily read from other interfaces.

Regarding alternative packages, `h5py` is straight-forward to use, but creates another dependence. `pandas` compatible HDF has some advantages within a Python environment, but the data structure is admittedly convoluted.

**Comment by speth** [comment]
Well, writing HDF5 files from `pandas` requires `pytables` to be installed. So we're going to have to depend on one HDF5 interface or another. I don't have a strong preference for either. I've used `h5py` more in the past and found it easy enough to use, but `pytables` seems like a reasonable option as well.

**Comment by ischoegl** [comment]
@speth ... I likewise have used `h5py` in the past and it *looks* simpler than `pytables` (which I haven't used directly), but that's not really a solid basis for a decision. Should the HDF package be added as a required dependency? Also, do you anticipate writing HDF from within C++ at some point in the future?

PS: I'd be happy to work on this once #805 is taken care of.

**Comment by bryanwweber** [comment]
I have used both in the past (although it has been some time) and h5py is (was) much easier to use for simple things, but is also somewhat less powerful for advanced use cases. I suspect that h5py will be sufficient for our purposes. As a longer-term idea, we could write our own wrapper API that only exposed the subset of functionality that we needed from either h5py or pytables. Then we could swap out the dependency if we wanted in the future (or possibly allow users to pick which one they want to use).

**Comment by ischoegl** [comment]
After some more work on #805 (which introduced HDF output/input for onedim), I believe that `h5py` will allow for adequate flexibility (while avoiding the added complexity of `pytables`). 

As a friendly amendment, I am proposing to introduce an intermediate layer. Assuming that the storage `key` (or `group` in HDF parlance) is 'somekey', the following would be an implementation for a `SolutionArray`:
```
somekey/d0/attrs['type'] = 'SolutionArray'
somekey/d0/attrs['mech'] = 'gri30.yaml'
somekey/d0/T
somekey/d0/density
somekey/d0/Y_H2
somekey/d0/Y_H
...
```
whereas for a fully documented `FlameBase` object it would be:
```
somekey/attrs['type`] = 'FreeFlame'
somekey/attrs[<setting1>] = ...
somekey/attrs[<setting2>] = ...
... 
somekey/d0/attrs['type'] = 'Inlet1D'
somekey/d0/attrs['mech'] = 'gri30.yaml'
somekey/d0/T
somekey/d0/density
somekey/d0/Y_H2
somekey/d0/Y_H
...
somekey/d1/attrs['type'] = 'Domain1D'
somekey/d1/attrs['mech'] = 'gri30.yaml'
somekey/d1/T
somekey/d1/density
somekey/d1/Y_H2
somekey/d1/Y_H
...
somekey/d2/attrs['type'] = 'Outlet1D'
somekey/d2/attrs['mech'] = 'gri30.yaml'
```
Any of the `dX` entries would constitute a well-formed (and readable) `SolutionArray` (even `Outlet1D` which would be empty). 

The `SolutionArray.read_hdf` reader would look for `SolutionArray` or `Domain1D` traces by default, with others being selectable by another optional flag. The structure could be easily extended for more user-friendly `ReactorNet` loggers.

**Comment by ischoegl** [comment]
@speth and @bryanwweber ... I implemented a draft version in #840. Still need to ~add `settings` and~ change some plumbing, but the bones are there. Feel free to comment.

One question I have is whether `h5py` should be introduced as a dependency.

**Comment by ischoegl** [comment]
I guess I never commented here that the suggested solution is now complete and ready for review - see #840.
